test/Dialect/FIRRTL/errors.mlir

```diff
@@ -871,3 +871,115 @@ firrtl.circuit "Foo"   {
+    %0 = firrtl.ref.resolve %ref_in1 : !firrtl.ref<uint<1>>
+  }
+}
+
+// -----
+// Incorrect connections
+
+firrtl.circuit "MyView" {
+  firrtl.module @MyView() {
+    %ref_in1 = firrtl.wire : !firrtl.ref<uint<1>>
+    %in1 = firrtl.wire : !firrtl.uint<1>
+    // expected-error @+1 {{that reference port operand can only be a module or instance port}}
+    firrtl.ref.send  %ref_in1, %in1 : !firrtl.ref<uint<1>>
+  }
+}
+
+// -----
+// Check upward reference XMRs
+
+firrtl.circuit "func2" {
+  firrtl.module @func2(in %ref_in1: !firrtl.ref<uint<1>>, out %ref_out: !firrtl.ref<uint<1>>) {
+    // expected-error @+1 {{connect has invalid flow for Ref type ports: the source expression "ref_in1" and destination expression "ref_out" both have same port kind, expected Module port to Instance connections only}}
+    firrtl.strictconnect %ref_out, %ref_in1 : !firrtl.ref<uint<1>>
+  }
+}
+
+// -----
+// Check flow semantics for ref.recv
+
+firrtl.circuit "Foo" {
+  // expected-note @+1 {{source was defined here}}
+  firrtl.module @Foo(out %_a: !firrtl.ref<uint<1>>) {
+    %a = firrtl.wire : !firrtl.uint<1>
+    // expected-error @+1 {{connect has invalid flow: the source expression "_a" has sink flow, expected source or duplex flow}}
+    firrtl.ref.recv %a, %_a : !firrtl.ref<uint<1>>
+  }
+}
+
+// -----
+// Check flow semantics for ref.send
+
+firrtl.circuit "Foo" {
+  // expected-note @+1 {{destination was defined here}}
+  firrtl.module @Foo(in  %_a: !firrtl.ref<uint<1>>) {
+    %a = firrtl.wire : !firrtl.uint<1>
+    // expected-error @+1 {{connect has invalid flow: the destination expression "_a" has source flow, expected sink or duplex flow}}
+    firrtl.ref.send %_a, %a : !firrtl.ref<uint<1>>
+  }
+}
+
+// -----
+// Check flow semantics for ref.recv
+
+firrtl.circuit "Bar" {
+  firrtl.module @Bar(in %_a: !firrtl.ref<uint<1>>) {
+    %a = firrtl.wire : !firrtl.uint<1>
+    %b = firrtl.wire : !firrtl.uint<1>
+    // expected-error @+1 {{ref recv (xmr write) result "a" cannot be used as the destination of a connect}}
+    firrtl.ref.recv %a, %_a : !firrtl.ref<uint<1>>
+    // expected-note @+1 {{the connect was defined here}}
+    firrtl.strictconnect %a, %b : !firrtl.uint<1>
+  }
+}
+
+// -----
+// Reference port cannot be reused
+
+firrtl.circuit "ForwardToInstance2x" {
+  firrtl.module @Bar2(out %_a: !firrtl.ref<uint<1>>) {
+    %zero = firrtl.constant 0 : !firrtl.uint<1>
+    firrtl.ref.send %_a, %zero : !firrtl.ref<uint<1>>
+  }
+  firrtl.module @Bar(out %_a: !firrtl.ref<uint<1>>) {
+    %x = firrtl.instance x @Bar2(out _a: !firrtl.ref<uint<1>>)
+    %y = firrtl.instance y @Bar2(out _a: !firrtl.ref<uint<1>>)
+    // expected-error @+1 {{connect operands of Ref type cannot be reused}}
+    firrtl.strictconnect %_a, %x : !firrtl.ref<uint<1>>
+    firrtl.strictconnect %_a, %y : !firrtl.ref<uint<1>>
```

Comment:
I grok that this is expedient, but I greatly dislike introducing "single connect semantics" as opposed to continuing to use "last connect semantics". I.e., this is creating an operation which is very "not FIRRTL".

Reply:
Ah, you are right, "last connect semantics" makes sense here. We can remove this constraint.